### PR TITLE
Feature/twitter oauth

### LINKED ISSUE #21: Twitter authentication no longer working
Cannot login to https://littleprinter.nordprojects.co/ by clicking "Sign in with Twitter" Receiving `OAuth error from twitter!` message:

image

### PR DESCRIPTION
**Pull request by phillipjf**
This MR switches Twitter login over to use [Twitter OAuth2.0](https://developer.twitter.com/en/docs/authentication/oauth-2-0) for fetching the username. This is primarily for historical purposes to continue using Twitter usernames for identifying users. Due to changes with the Twitter API, the friends functionality is no longer feasible to maintain and was not exposed in the UI anyways. Flask-Dance project even removed Twitter in a recent release [Flask-Dance#416](https://github.com/singingwolfboy/flask-dance/pull/416)

**BREAKING:** This MR includes a migration to remove the friends association in the DB.

Resolves #21.

### CODE REVIEW

**Comment by phillipjf** [comment]
@joerick I think this covers everything to resolve #21. Will need to get the OAuth 2.0 Client ID and Client Secret from the Twitter Developer Portal and set the new environment variables (`TWITTER_OAUTH_CLIENT_KEY`, `TWITTER_OAUTH_CLIENT_SECRET`, `OAUTH_REDIRECT_URI`) accordingly.

File: migrations/versions/89a031bd9657_remove_friends.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_unique_constraint(None, "claim_code", ["claim_code"])
+    op.create_foreign_key(None, "print_key", "print_key", ["parent_id"], ["id"])
+    op.create_unique_constraint(None, "printer", ["used_claim_code"])
```
Comment (joerick) [inline migrations/versions/89a031bd9657_remove_friends.py]: not sure why alembic inserted these, but they don't look related so I'm inclined to remove them
```suggestion
```

File: migrations/versions/89a031bd9657_remove_friends.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        sa.Column(
+            "last_friend_refresh",
+            postgresql.TIMESTAMP(),
+            autoincrement=False,
+            nullable=True,
+        ),
+    )
+    op.drop_constraint(None, "printer", type_="unique")
+    op.drop_constraint(None, "print_key", type_="foreignkey")
+    op.drop_constraint(None, "claim_code", type_="unique")
```
Comment (joerick) [inline migrations/versions/89a031bd9657_remove_friends.py]:
```suggestion
```
likewise here

File: sirius/web/twitter.py [hunk 3]
```diff
@@ -1,51 +1,75 @@
...
 logger = logging.getLogger(__name__)
 
-api_key=os.environ.get('TWITTER_CONSUMER_KEY', 'DdrpQ1uqKuQouwbCsC6OMA4oF')
-api_secret=os.environ.get('TWITTER_CONSUMER_SECRET', 'S8XGuhptJ8QIJVmSuIk7k8wv3ULUfMiCh9x1b19PmKSsBh1VDM')
+blueprint = flask.Blueprint("twitter", __name__)
+
+auth_url = "https://twitter.com/i/oauth2/authorize"
+token_url = "https://api.twitter.com/2/oauth2/token"
+
+scopes = ["tweet.read", "users.read", "offline.access"]
```
Comment (joerick) [inline sirius/web/twitter.py:18]: Do we need `tweet,read`? I also wonder about `users.read`?
```suggestion
scopes = ["users.read", "offline.access"]
```

File: Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.7.4-slim-buster
+FROM python:3.7-slim-bullseye
```
Comment (joerick) [inline Dockerfile]: I had to roll this back to get the CI to run -
```suggestion
FROM python:3.7.4-slim-buster
```

File: sirius/web/twitter.py (hunk 3 above)
Comment (phillipjf) [reply sirius/web/twitter.py:18]: These are the scopes listed under the docs for `GET /2/users/me`:

https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-me

File: Dockerfile (hunk 4 above)
Comment (phillipjf) [reply Dockerfile]: Interesting. I was having trouble resolving packages on `buster` so I bumped to `bullseye` so it's not a big deal to me.

Comment (joerick) [reply Dockerfile]: Ah, might be cos I'm testing on my arm64 mac, I'll try it with x64

**Comment by joerick** [comment]
If you have a chance to look at why CI is failing, that would be much appreciated! Something dependency-based. 

By the way, another thought I had was... the site is hosted on Heroku, it uses their buildpacks, not the Docker config. Do you know if the switch to pyproject.toml will work other there?

**Comment by phillipjf** [comment]
> If you have a chance to look at why CI is failing, that would be much appreciated! Something dependency-based.

Fixed this, the reference to the `snapshottest` package was the culprit. Also a couple tests needed updating.

> By the way, another thought I had was... the site is hosted on Heroku, it uses their buildpacks, not the Docker config. Do you know if the switch to pyproject.toml will work other there?

I _think_ this will work with the empty `setup.py` as recommended here: https://github.com/heroku/heroku-buildpack-python/pull/834#issuecomment-1242217467

File: Dockerfile (hunk 4 above)
Comment (phillipjf) [reply Dockerfile]: I am too so maybe it was just a transient issue. I was also testing some other changes (using `imgkit` instead of `selenium`/`phantomjs`) so it's probably just something I misinterpreted.

**Comment by joerick** [comment]
Brilliant! Lets give this a try...

**Comment by joerick** [comment]
It works!

Screenshot 2023-07-07 at 3 31 46 pm
Screenshot 2023-07-07 at 3 32 00 pm

**Comment by joerick** [comment]
Awesome. Thank you so much @phillipjf for this! Once I had all the API creds in the right place, it worked flawlessly.
